### PR TITLE
QA-3004 base_utility.py: add 'match_once_or_more' option + refactoring

### PR DESCRIPTION
**Pull request by iartiukhov**

### CODE REVIEW

File: src/tiden/utilities/base_utility.py [hunk 1]
```diff
@@ -67,19 +68,44 @@ def check_content_all_required(self, buff, lines_to_search, maintain_order=False
...
+                )
             else:
-                raise TidenException('Searching \n%s \nand found: \n%s \nin buffer \n%s.\nFound additional items:\n%s'
-                                     % ('\n'.join(search_for),
-                                        '\n'.join(found),
-                                        '\n'.join(search_in),
-                                        set(found).difference(set(search_for))))
+                # len(search_for) < len(found)
+                if not match_once_or_more or \
+                        (match_once_or_more and (len(search_for_uniq) != len(found_uniq))):
```
Comment (MaxLosevskoy) [inline src/tiden/utilities/base_utility.py:102]:
```suggestion
                if not match_once_or_more or len(search_for_uniq) != len(found_uniq):
```

Comment (iartiukhov) [reply src/tiden/utilities/base_utility.py]: Applied, thanks!

**Comment by iartiukhov** [comment]
Closing for improvements

**Comment by mshonichev** [comment]
Please, provide self-contained unit tests for this function, cause it's really a core method.

File: tests/test_baseline_utility.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    lines_to_search = ["hello"]
+    assert util.check_content_all_required(buff, lines_to_search)
+
+
+def test_simple_negative():
+    with pytest.raises(TidenException):
+        util = BaseUtility(MockIgnite())
+        buff = "hola!\nhello!\nhi!"
+        lines_to_search = ["ni hao!"]
+        util.check_content_all_required(buff, lines_to_search)
```
Comment (MaxLosevskoy) [inline tests/test_baseline_utility.py:23]:
```suggestion


def test_simple_negative():
    with pytest.raises(TidenException):
        util = BaseUtility(MockIgnite())
        buff = "hola!\nhello!\nhi!"
        lines_to_search = ["ni hao!"]
        util.check_content_all_required(buff, lines_to_search)


def test_partial_match():
    util = BaseUtility(MockIgnite())
    buff = "hola!\nhello!\nhi!"
    lines_to_search = ["hell"]
    assert util.check_content_all_required(buff, lines_to_search)


def test_multiple_lines():
    util = BaseUtility(MockIgnite())
    buff = "hola!\nhello!\nhi!"
    lines_to_search = ["hi!", "hola!"]
    assert util.check_content_all_required(buff, lines_to_search)


def test_repeated_line():
    util = BaseUtility(MockIgnite())
    buff = "hola!\nhello!\nhi!"
    lines_to_search = ["hi!", "hi!"]
    assert util.check_content_all_required(buff, lines_to_search)


def test_multiple_lines_negative():
    with pytest.raises(TidenException):
        util = BaseUtility(MockIgnite())
        buff = "hola!\nhello!\nhi!"
        lines_to_search = ["hola!", "ni hao!"]
        util.check_content_all_required(buff, lines_to_search)


def test_spaces_negative():
    with pytest.raises(TidenException):
        util = BaseUtility(MockIgnite())
        buff = "hola!\nhello!\nhi!"
        lines_to_search = [" hi!\n"]
        util.check_content_all_required(buff, lines_to_search)
```

Comment (MaxLosevskoy) [reply tests/test_baseline_utility.py:23]: Regarding the last test - maybe it'd be convenient to trim the leading and trailing spaces/new lines, then the test becomes positive.

Maybe the whole suggestion is a junk :slightly_smiling_face:

File: src/tiden/utilities/base_utility.py [hunk 3]
```diff
@@ -27,59 +27,94 @@ def __init__(self, ignite, parent_cls=None):
...
-            if escape_from_search:
-                search_in = [item for item in search_in if item not in escape_from_search]
-
-        search_for = list(lines_to_search)
-        found = []
-        result = True
+        if isinstance(lines_to_search, list):
+            lines_to_search_0 = lines_to_search
+        else:
+            lines_to_search_0 = list(lines_to_search)
```
Comment (MaxLosevskoy) [inline src/tiden/utilities/base_utility.py:49]: lines_to_search_0 = list(lines_to_search)

Comment (iartiukhov) [reply src/tiden/utilities/base_utility.py:49]: `lines_to_search_0 = list(lines_to_search)` is wrong logic. Thanks to pointing this out :)

Comment (MaxLosevskoy) [reply src/tiden/utilities/base_utility.py]: Maybe, I just simplified keeping the semantics)

Comment (iartiukhov) [reply src/tiden/utilities/base_utility.py]: The semantics was wrong

File: tests/test_baseline_utility.py (hunk 2 above)
Comment (iartiukhov) [reply tests/test_baseline_utility.py:23]: @MaxLosevskoy Thanks for useful cases!

**Comment by iartiukhov** [comment]
> Please, provide self-contained unit tests for this function, cause it's really a core method.

@mshonichev I've added some tests. Please review.

File: src/tiden/utilities/base_utility.py [hunk 4]
```diff
@@ -27,59 +27,94 @@ def __init__(self, ignite, parent_cls=None):
...
-            if escape_from_search:
-                search_in = [item for item in search_in if item not in escape_from_search]
-
-        search_for = list(lines_to_search)
-        found = []
-        result = True
+        if isinstance(lines_to_search, list):
+            lines_to_search_0 = lines_to_search
+        else:
+            lines_to_search_0 = [lines_to_search]
```
Comment (MaxLosevskoy) [inline src/tiden/utilities/base_utility.py:49]: Do we need to distinguish `lines_to_search` from `lines_to_search_0`?
Maybe we can simply put
```py
if not isinstance(lines_to_search, list):
    lines_to_search = [lines_to_search]
```
?

File: src/tiden/utilities/base_utility.py [hunk 5]
```diff
@@ -27,59 +27,94 @@ def __init__(self, ignite, parent_cls=None):
...
         else:
+            search_for = lines_to_search_0
+
+        found = []
+        match_found = False
+
+        # Iterate through all input lines
+        for line_in in search_in:
+            # Skip current line if it matches any line from the "escape" list
+            if escape and any(map(lambda esc_line: esc_line in line_in, escape)):
```
Comment (MaxLosevskoy) [inline src/tiden/utilities/base_utility.py:63]: Why not just
```py
if escape and line_in in escape:
```
?
Am I correct that items in `escape` is not just exact lines to skip, but also a substrings of such lines?

File: src/tiden/utilities/base_utility.py (hunk 4 above)
Comment (iartiukhov) [reply src/tiden/utilities/base_utility.py]: @MaxLosevskoy It turns out, that we can do this. Fixed.

File: src/tiden/utilities/base_utility.py (hunk 5 above)
Comment (iartiukhov) [reply src/tiden/utilities/base_utility.py:61]: You cannot do this. `escape` is a substring of a whole line.

**Review by MaxLosevskoy — APPROVED** [review]

**Comment by mshonichev** [comment]
Finally, that's how one need to improve framework! Thank you for great contribution!

**Pull request merged**